### ISSUE TITLE
Building Kubevirt for ppc64le 

### ISSUE THREAD

**Issue by danielhb** [kubevirt/kubevirt#4427]
Hi,

I'm trying to run Kubevirt in a ppc64 cluster. The intent is to do a sanity check of the
current upstream code to see if there is something to be fixed in that platform. The
idea is to re-enable ppc64 support quicker when the COPR ppc64 builds are sorted out.
I can have the changes ready in my personal repo, then contribute it back when ppc64le
is back to business in Kubevirt.

I created a Libvirt ppc64le image using the kubevirt/libvirt repo (using an amd64 host). Back in the
kubevirt repo, in a Power 8 server running Fedora 32, I made the adjustments to re-enable ppc64le
(pointed to a valid libvirt_ppc64le image, re-enabled ppc64le in hack/builder/version.sh) and, when
attempting to execute 'make' I got a 'format exec error' error when executing the 'hack/dockerized' script.

I realized that I should have a kubevirt-builder image for ppc64le as well, so  I created one (again
using an amd64 box)  and used it as KUBEVIRT_BUILDER_IMAGE back in the P8 host. I'm still not
able to execute a 'make'  there - it errors out right before "Run the command", somewhere inside
hack/bazel-fmt.sh (still debugging to see the reason).

Meanwhile, I tried to go back to amd64 and build Kubevirt with ppc64le support enabled. Aside
from the KUBEVIRT_BUILDER_IMAGE change I replicated all the changes I did in the Power host. It
went further than my attempt in Power, but I don't see any images uploaded in the "DOCKER_PREFIX"
repo.

What am I missing/doing wrong? Can I build kubevirt for ppc64le using a Power host or should I do
everything in an amd64 env? Do I need more changes than the obvious  one such as add "ppc64le" in
the ARCHITECTURES var and pointing to the right Libvirt container image?

Any pointer is appreciated. Thanks!

**Comment by maya-r** [comment]
cc @murilofv if you haven't seen this yet

**Comment by danielhb** [comment]
I managed to build and run Kubevirt for ppc64le building all artifacts. Shout out to @murilofv and @andreabolognani for the assistance and patience answering all my questions. For future reference I'll sum up what I ended up doing to make it happen.

1 - build kubevirt/libvirt for ppc64le using a x86 env. 

The Libvirt container images are built using https://github.com/kubevirt/libvirt. To see what I did to make it work check out my mirror https://github.com/danielhb/libvirt-2/tree/ppc64_build. The scripts are wired for cross compiling using 'docker buildx' in x86, so just use a x86 box to build it.

In hack/config change 'IMAGE_NAME' in to point to your Docker repo of choice and Libvirt/QEMU versions. TARGET_ARCHITECTURE needs to include "ppc64le" as well. Change Dockerfile.in if necessary - in my case COPR isn't back for ppc64le so I had to generated the packages, copy them to the container and install. "hack/prepare.sh && hack/build.sh && hack/deploy.sh"  will generate and publish a Libvirt container to the Docker repo pointed in IMAGE_NAME.

2 - build kubevirt-builder for ppc64 using a x86 env

For the same reason as (1) it's easier to just use a x86 env to build kubevirt-builder for ppc64le. The changes here are all about fixing ppc64le package references and so on. It's easier to just go to my mirror  in https://github.com/danielhb/kubevirt/tree/ppc64_builder_build_v2 to see what I had to change.

"make builder-build && make builder-publish" will hopeffuly get you a kubevirt-builder image.

3 - in a ppc64le host, build kubevirt

Finally, point to the Libvirt container image you built in step 1 in the WORKSPACE file, rule 'libvirt_ppc64le'. Also,
point to the kubevirt-builder container image from step 2 in the hack/dockerized file, variable KUBEVIRT_BUILDER_IMAGE. 

Set DOCKER_PREFIX to the Docker repo you want Kubevirt images to be published. Then execute

"make bazel-build-images && make bazel-push-images"

If everything went fine, then it's just a matter of executing 'make' and deploying the manifests via 'kubectl create -f' like it is done in the documentation.

NOTE: 'make cluster-up' does not work out of the box for ppc64le. You'll either want to hack it to make it work for your arch or you'll need a pre-configured cluster to deploy via "kubectl create -f". I chose the latter.


I'll close the bug to make it clear that this is not a pending issue. Thanks!

**Comment by andreabolognani** [comment]
> I managed to build and run Kubevirt for ppc64le building all artifacts. Shout out to @murilofv and @andreabolognani for the assistance and patience answering all my questions. For future reference I'll sum up what I ended up doing to make it happen.
> 
> 1 - build kubevirt/libvirt for ppc64le using a x86 env.
> 
> The Libvirt container images are built using https://github.com/kubevirt/libvirt. To see what I did to make it work check out my mirror https://github.com/danielhb/libvirt-2/tree/ppc64_build. The scripts are wired for cross compiling using 'docker buildx' in x86, so just use a x86 box to build it.
> 
> In hack/config change 'IMAGE_NAME' in to point to your Docker repo of choice and Libvirt/QEMU versions. TARGET_ARCHITECTURE needs to include "ppc64le" as well. Change Dockerfile.in if necessary - in my case COPR isn't back for ppc64le so I had to generated the packages, copy them to the container and install. "hack/prepare.sh && hack/build.sh && hack/deploy.sh" will generate and publish a Libvirt container to the Docker repo pointed in IMAGE_NAME.

Yeah, unfortunately that is still a work in progress on the COPR side, with no ETA that I'm aware of, so not much chance of making this anything other than a hack for the time being :(

> 2 - build kubevirt-builder for ppc64 using a x86 env
> 
> For the same reason as (1) it's easier to just use a x86 env to build kubevirt-builder for ppc64le. The changes here are all about fixing ppc64le package references and so on. It's easier to just go to my mirror in https://github.com/danielhb/kubevirt/tree/ppc64_builder_build_v2 to see what I had to change.
> 
> "make builder-build && make builder-publish" will hopeffuly get you a kubevirt-builder image.

I skimmed your changes and it looks to me like some of them could be upstreamed right away, for example reintroducing the ppc64le packages.

Making the `hack/dockerized` script multi-arch aware is also something that is going to be important if we ever hope to have ppc64le as a first-class citizen again, so you could prepare for that and get the patches upstream even if a ppc64le `kubevirt/builder` image can't be built properly at the moment.

Speaking of which, it looks like the only real blocker for that happening is bazel binary releases not being available for ppc64le the same way they are for x86_64 and aarch64? I think we need to bug the bazel developers if we want to have any hope that will change going forward...

> 3 - in a ppc64le host, build kubevirt
> 
> Finally, point to the Libvirt container image you built in step 1 in the WORKSPACE file, rule 'libvirt_ppc64le'. Also,
> point to the kubevirt-builder container image from step 2 in the hack/dockerized file, variable KUBEVIRT_BUILDER_IMAGE.
> 
> Set DOCKER_PREFIX to the Docker repo you want Kubevirt images to be published. Then execute
> 
> "make bazel-build-images && make bazel-push-images"
> 
> If everything went fine, then it's just a matter of executing 'make' and deploying the manifests via 'kubectl create -f' like it is done in the documentation.
> 
> NOTE: 'make cluster-up' does not work out of the box for ppc64le. You'll either want to hack it to make it work for your arch or you'll need a pre-configured cluster to deploy via "kubectl create -f". I chose the latter.

Fixing `kubevirt/kubevirtci` so that `make cluster-up` works as smoothly on ppc64le as it does on x86_64 seems like another obvious task that can be worked on even while COPR can't produce ppc64le RPMS.

> I'll close the bug to make it clear that this is not a pending issue. Thanks!

I'm not sure I agree with this decision - it's certainly not possible to build KubeVirt on ppc64le without really heavy hacking right now, and I see nothing wrong with tracking this fact with an issue...

**Comment by danielhb** [comment]
> > "make builder-build && make builder-publish" will hopeffuly get you a kubevirt-builder image.
> 
> I skimmed your changes and it looks to me like some of them could be upstreamed right away, for example reintroducing the ppc64le packages.

Note that I didn't generate the google storageapi link for the ppc64 packages (I have no idea how to do that hehe).

> 
> Making the `hack/dockerized` script multi-arch aware is also something that is going to be important if we ever hope to have ppc64le as a first-class citizen again, so you could prepare for that and get the patches upstream even if a ppc64le `kubevirt/builder` image can't be built properly at the moment.

Makes sense. Guess we can make the script architecture aware and pick an appropriate kubevirt-builder image for the arch or something similar.

> 
> Speaking of which, it looks like the only real blocker for that happening is bazel binary releases not being available for ppc64le the same way they are for x86_64 and aarch64? I think we need to bug the bazel developers if we want to have any hope that will change going forward...

TBH I don't understand the Bazel situation. The package compiles in ppc64le without too much problems if you follow

https://github.com/vbatts/copr-build-bazel

but Bazel ppc64le packages aren't distributed in the distros ever since RHEL 7.7. 


> 
> > 3 - in a ppc64le host, build kubevirt
> > Finally, point to the Libvirt container image you built in step 1 in the WORKSPACE file, rule 'libvirt_ppc64le'. Also,
> > point to the kubevirt-builder container image from step 2 in the hack/dockerized file, variable KUBEVIRT_BUILDER_IMAGE.
> > Set DOCKER_PREFIX to the Docker repo you want Kubevirt images to be published. Then execute
> > "make bazel-build-images && make bazel-push-images"
> > If everything went fine, then it's just a matter of executing 'make' and deploying the manifests via 'kubectl create -f' like it is done in the documentation.
> > NOTE: 'make cluster-up' does not work out of the box for ppc64le. You'll either want to hack it to make it work for your arch or you'll need a pre-configured cluster to deploy via "kubectl create -f". I chose the latter.
> 
> Fixing `kubevirt/kubevirtci` so that `make cluster-up` works as smoothly on ppc64le as it does on x86_64 seems like another obvious task that can be worked on even while COPR can't produce ppc64le RPMS.

Indeed it can. 

> 
> > I'll close the bug to make it clear that this is not a pending issue. Thanks!
> 
> I'm not sure I agree with this decision - it's certainly not possible to build KubeVirt on ppc64le without really heavy hacking right now, and I see nothing wrong with tracking this fact with an issue...


The idea of this bug was to ask for assistance, but I don't see any problems using it as tracker for ppc64le support for Kubevirt. I'll reopen it

**Comment by murilofv** [comment]
> > 2 - build kubevirt-builder for ppc64 using a x86 env
> > For the same reason as (1) it's easier to just use a x86 env to build kubevirt-builder for ppc64le. The changes here are all about fixing ppc64le package references and so on. It's easier to just go to my mirror in https://github.com/danielhb/kubevirt/tree/ppc64_builder_build_v2 to see what I had to change.
> > "make builder-build && make builder-publish" will hopeffuly get you a kubevirt-builder image.
> 
> I skimmed your changes and it looks to me like some of them could be upstreamed right away, for example reintroducing the ppc64le packages.
> 
> Making the `hack/dockerized` script multi-arch aware is also something that is going to be important if we ever hope to have ppc64le as a first-class citizen again, so you could prepare for that and get the patches upstream even if a ppc64le `kubevirt/builder` image can't be built properly at the moment.
> 
@andreabolognani Can you go a bit into more detail about what you think needs to be changed? My understanding is that given the infrastructure that Roman created in this commit https://github.com/kubevirt/kubevirt/commit/a029feaa2ae6d25cef3458db8a188cc350c9f0de it should be just a matter of pointing to the manifest that contains both the builder for amd64 and ppc64le and docker would automatically pull and use the appropriate one for the architecture, that way no additional changes would be needed to this particular script.

>Speaking of which, it looks like the only real blocker for that happening is bazel binary releases not being available for ppc64le the same way they are for x86_64 and aarch64? I think we need to bug the bazel developers if we want to have any hope that will change going forward...

Yeah, in the past Kubevirt picked up bazel from vbatts COPR but with the COPR issues mentioned the ppc64le builder was affected at the time, but now it looks like Kubevirt changed the location of where it gets the bazel binary in this commit https://github.com/kubevirt/kubevirt/commit/6618cf637e5499ce66a32603fa171acca2c4c4e7 and now it is as you mentioned, we need to reach out to the bazel developers to get this sorted out for the POWER platform.

**Comment by andreabolognani** [comment]
> > Making the `hack/dockerized` script multi-arch aware is also something that is going to be important if we ever hope to have ppc64le as a first-class citizen again, so you could prepare for that and get the patches upstream even if a ppc64le `kubevirt/builder` image can't be built properly at the moment.
> 
> @andreabolognani Can you go a bit into more detail about what you think needs to be changed? My understanding is that given the infrastructure that Roman created in this commit [a029fea](https://github.com/kubevirt/kubevirt/commit/a029feaa2ae6d25cef3458db8a188cc350c9f0de) it should be just a matter of pointing to the manifest that contains both the builder for amd64 and ppc64le and docker would automatically pull and use the appropriate one for the architecture, that way no additional changes would be needed to this particular script.

I mistakenly thought that the sha256 stored in `KUBEVIRT_BUILDER_IMAGE` was pointing to a single-architecture image, but I see now that's not actually the case:

```
$ docker manifest inspect kubevirt/builder@sha256:5e0c047e3cac6123419c972bb9d31d89f686751b962190c147379d846f5d1e46
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 2216,
         "digest": "sha256:af6539124956056ce053d99afd8f5eeec318bd725300f952edb1de5f81f0d491",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 2216,
         "digest": "sha256:0de09c0fe9bff0aa62d50f83451bbc979476f0165d4fe6b558440007d15925d0",
         "platform": {
            "architecture": "ppc64le",
            "os": "linux"
         }
      }
   ]
}
```

So you're absolutely correct when you say that no further changes to that script are necessary :)

> > Speaking of which, it looks like the only real blocker for that happening is bazel binary releases not being available for ppc64le the same way they are for x86_64 and aarch64? I think we need to bug the bazel developers if we want to have any hope that will change going forward...
> 
> Yeah, in the past Kubevirt picked up bazel from vbatts COPR but with the COPR issues mentioned the ppc64le builder was affected at the time, but now it looks like Kubevirt changed the location of where it gets the bazel binary in this commit [6618cf6](https://github.com/kubevirt/kubevirt/commit/6618cf637e5499ce66a32603fa171acca2c4c4e7) and now it is as you mentioned, we need to reach out to the bazel developers to get this sorted out for the POWER platform.

I guess the other way to look at it is that, once COPR can once again build packages for ppc64le, we'll be able solve both the libvirt issue *and* the bazel issue by switching back to the `vbatts/bazel` COPR :)

tl;dr `cluster-up` could use some work, but most of the other blockers for ppc64le support are going to be magically solved once COPR has been fixed.

**Comment by andreabolognani** [comment]
@danielhb in the months since this issue was created, KubeVirt has gained aarch64 support.

This support is solid enough that cross-builds are run by default for all PRs (see [here](https://github.com/kubevirt/kubevirt/pull/5328#issuecomment-830149969) for an example of cross-building issues blocking a change).

So I think that, at this point, someone who wanted to make KubeVirt available on ppc64le would basically just need to look at how aarch64 support has been implemented and copy it :) And perhaps keeping this issue around no longer makes sense. How do you feel about that?

Of course the underlying issue is still getting COPR to produce ppc64le builds... Do you have any updates on that situation?

**Comment by danielhb** [comment]
@andreabolognani since the ARM support apparently cut our work short, we can close this issue in my opinion.

WRT COPR, I asked around and apparently the situation is still the same, unfortunately. The Fedora tech support staff has a lot of work to do, and getting the ppc64le builders online aren't their top priority ATM.

In fact, COPR online for ppc64le is a hard requirement for ppc64le support in Kubevirt. We need those servers online to be able to reliably get the ppc64le Libvirt container images and so on. Trying to work around it by custom created images, like I did here (https://github.com/kubevirt/kubevirt/issues/4427#issuecomment-725327572) works for some time. Then upstream keeps going forward and eventually, unless you're willing to keep all the custom images up to date for each release, your support will become  obsolete. We're way better off waiting for the COPR situation to be solved first.
